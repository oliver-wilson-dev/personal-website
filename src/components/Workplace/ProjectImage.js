import React, { useState, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './index.css';

/*
  Setting the image src when rendering on the client because on the server
  no onerror event is respected.

  https://stackoverflow.com/questions/51639891/detect-broken-image-when-using-hydrate-instead-of-render-from-react-dom
*/

const ProjectImage = ({ alt, src: srcProp, linkToProject }) => {
  const [imgLoadFailed, setImgLoadFailed] = useState(false);
  const [src, setSrc] = useState(undefined);

  useEffect(() => setSrc(srcProp), []);

  const onError = useCallback(() => setImgLoadFailed(true), [setImgLoadFailed]);

  if (imgLoadFailed) return null;

  return (
    <div
      className={styles.projectImageWrapper}
    >
      <a href={linkToProject}>
        <img
          alt={alt}
          src={src}
          loading="lazy"
          onError={onError}
          className={styles.image}
        />
      </a>
    </div>
  );
};

ProjectImage.propTypes = {
  alt: PropTypes.string,
  src: PropTypes.string,
  linkToProject: PropTypes.string,
};

export default ProjectImage;
