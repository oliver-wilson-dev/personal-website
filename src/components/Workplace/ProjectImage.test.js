import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import ProjectImage from './ProjectImage';

const defaultProps = {
  alt: 'test-alt',
  src: 'test-src',
  linkToProject: 'test-linkToProject',
};


const render = ({ props = {}, renderMethod = shallow } =
{ props: {}, renderMethod: shallow }) => renderMethod(
  <ProjectImage {...defaultProps} {...props} />
);

describe('ProjectImage', () => {
  it('should render correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  it('should not set a src property', () => {
    expect(render().find('img').prop('src')).toBe(undefined);
  });

  describe('when mounted on the client', () => {
    it('should render correctly', () => {
      expect(render({ renderMethod: mount })).toMatchSnapshot();
    });

    it('should set the src property', () => {
      const src = 'test-src';
      expect(render({ renderMethod: mount, props: { src } }).find('img').prop('src'))
        .toBe(src);
    });

    describe('when the onError callback is triggered', () => {
      it('should render correctly', () => {
        const component = render({ renderMethod: mount });

        act(() => {
          component.find('img').prop('onError')();
        });

        component.update();
        expect(component).toMatchSnapshot();
      });
    });
  });
});
