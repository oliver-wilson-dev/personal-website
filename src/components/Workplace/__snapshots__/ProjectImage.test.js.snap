// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectImage should render correctly 1`] = `
<div
  className="projectImageWrapper"
>
  <a
    href="test-linkToProject"
  >
    <img
      alt="test-alt"
      className="image"
      loading="lazy"
      onError={[Function]}
    />
  </a>
</div>
`;

exports[`ProjectImage when mounted on the client should render correctly 1`] = `
<ProjectImage
  alt="test-alt"
  linkToProject="test-linkToProject"
  src="test-src"
>
  <div
    className="projectImageWrapper"
  >
    <a
      href="test-linkToProject"
    >
      <img
        alt="test-alt"
        className="image"
        loading="lazy"
        onError={[Function]}
        src="test-src"
      />
    </a>
  </div>
</ProjectImage>
`;

exports[`ProjectImage when mounted on the client when the onError callback is triggered should render correctly 1`] = `
<ProjectImage
  alt="test-alt"
  linkToProject="test-linkToProject"
  src="test-src"
/>
`;
